performance_manager/lib/l1_full_trip_events.py

```diff
@@ -0,0 +1,223 @@
+            VehiclePositionEvents.stop_id,
+            VehiclePositionEvents.direction_id,
+            VehiclePositionEvents.route_id,
+            VehiclePositionEvents.start_date,
+            VehiclePositionEvents.start_time,
+            VehiclePositionEvents.vehicle_id,
+        ).where(
+            (VehiclePositionEvents.is_moving == sa.false())
+            & (
+                VehiclePositionEvents.pk_id.notin_(
+                    sa.select(FullTripEvents.fk_vp_stopped_event)
+                )
+            )
+            & (VehiclePositionEvents.hash.notin_(dupe_hash_query))
+        ),
+        fk_rename_field="fk_vp_stopped_event",
+    )
+
+    stopped_tu_events = EventsToMerge(
+        select_query=sa.select(
+            TripUpdateEvents.pk_id,
+            TripUpdateEvents.stop_sequence,
+            TripUpdateEvents.stop_id,
+            TripUpdateEvents.direction_id,
+            TripUpdateEvents.route_id,
+            TripUpdateEvents.start_date,
+            TripUpdateEvents.start_time,
+            TripUpdateEvents.vehicle_id,
+        ).where(
+            (
+                TripUpdateEvents.pk_id.notin_(
+                    sa.select(FullTripEvents.fk_tu_stopped_event)
+                )
+            )
+        ),
+        fk_rename_field="fk_tu_stopped_event",
+    )
+
+    return {
+        "moving_vp": moving_vp_events,
+        "stopped_vp": stopped_vp_events,
+        "stopped_tu": stopped_tu_events,
+    }
+
+
+def load_and_transform(
+    events_list: Dict[str, EventsToMerge], db_manager: DatabaseManager
+) -> None:
+    """
+    load and transform each set of events from RDS L0 tables
+    """
+    expected_hash_columns = [
+        "stop_sequence",
+        "stop_id",
+        "direction_id",
+        "route_id",
+        "start_date",
+        "start_time",
+        "vehicle_id",
+    ]
+
+    for event_type in events_list.values():
+        event_type.merge_dataframe = db_manager.select_as_dataframe(
+            event_type.select_query
+        )
+
+        if event_type.merge_dataframe.shape[0] == 0:
+            event_type.merge_dataframe = pandas.DataFrame(
+                columns=["pk_id", "hash"]
+            )
+        else:
+            event_type.merge_dataframe = add_event_hash_column(
+                event_type.merge_dataframe,
+                expected_hash_columns=expected_hash_columns,
+            )[["pk_id", "hash"]]
+
+        event_type.merge_dataframe.rename(
+            columns={"pk_id": event_type.fk_rename_field}, inplace=True
+        )
+
+
+def merge_events(events_list: Dict[str, EventsToMerge]) -> pandas.DataFrame:
+    """
+    merge all unprocessed event types into single dataframe
+    """
+    return_dataframe = events_list["moving_vp"].merge_dataframe.merge(
+        events_list["stopped_vp"].merge_dataframe, how="outer"
+    )
+    return_dataframe = return_dataframe.merge(
+        events_list["stopped_tu"].merge_dataframe, how="left"
+    )
```

Comment:
why do we have the differing `outer` and `left` merge strategies here?

Reply:
The `outer` merge makes sure that we capture all vehicle_position events, since they are valid real-world occurrences. 

I'm just doing a left merge on the trip_update stop events because these are technically predictions, so the left merge drops any trip_update predictions without matching vehicle_position events.

I can add some comments clarifying these operations.